ledger/complete/wal/checkpoint_v6_reader.go

```diff
@@ -0,0 +1,722 @@
+	paths := make([]string, 0)
+
+	paths = append(paths, filePathCheckpointHeader(dir, fileName))
+
+	subtrieCount := subtrieCountByLevel(subtrieLevel)
+	for i := 0; i < subtrieCount; i++ {
+		paths = append(paths, path.Join(dir, fmt.Sprintf("%v.%v", fileName, i)))
+	}
+
+	p, _ := filePathTopTries(dir, fileName)
+	paths = append(paths, p)
+	return paths
+}
+
+// readCheckpointHeader takes a file path and returns subtrieChecksums and topTrieChecksum
+// any error returned are exceptions
+func readCheckpointHeader(filepath string, logger *zerolog.Logger) ([]uint32, uint32, error) {
+	closable, err := os.Open(filepath)
+	if err != nil {
+		return nil, 0, fmt.Errorf("could not seek to start for header file: %w", err)
+	}
+
+	var bufReader io.Reader = bufio.NewReaderSize(closable, defaultBufioReadSize)
+	reader := NewCRC32Reader(bufReader)
+	defer func(f *os.File) {
+		evictErr := evictFileFromLinuxPageCache(f, false, logger)
+		if evictErr != nil {
+			logger.Warn().Msgf("failed to evict header file %s from Linux page cache: %s", filepath, evictErr)
+			// No need to return this error because it's possible to continue normal operations.
+		}
+		f.Close()
+	}(closable)
+
+	// read the magic bytes and check version
+	err = validateFileHeader(MagicBytesCheckpointHeader, VersionV6, reader)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// read the subtrie level
+	subtrieCount, err := readSubtrieCount(reader)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	subtrieChecksums := make([]uint32, subtrieCount)
+	for i := uint16(0); i < subtrieCount; i++ {
+		sum, err := readCRC32Sum(reader)
+		if err != nil {
+			return nil, 0, fmt.Errorf("could not read %v-th subtrie checksum from checkpoint header: %w", i, err)
+		}
+		subtrieChecksums[i] = sum
+	}
+
+	// read top level trie checksum
+	topTrieChecksum, err := readCRC32Sum(reader)
+	if err != nil {
+		return nil, 0, fmt.Errorf("could not read checkpoint top level trie checksum in chechpoint summary: %w", err)
+	}
+
+	// calculate the actual checksum
+	actualSum := reader.Crc32()
+
+	// read the stored checksum, and compare with the actual sum
+	expectedSum, err := readCRC32Sum(reader)
+	if err != nil {
+		return nil, 0, fmt.Errorf("could not read checkpoint header checksum: %w", err)
+	}
+
+	if actualSum != expectedSum {
+		return nil, 0, fmt.Errorf("invalid checksum in checkpoint header, expected %v, actual %v",
+			expectedSum, actualSum)
+	}
+
+	err = ensureReachedEOF(reader)
+	if err != nil {
+		return nil, 0, fmt.Errorf("fail to read checkpoint header file: %w", err)
+	}
+
+	return subtrieChecksums, topTrieChecksum, nil
+}
+
+// allPartFileExist check if all the part files of the checkpoint file exist
+// it returns nil if all files exist
+// it returns os.ErrNotExist if some file is missing, use (os.IsNotExist to check)
+// it returns err if running into any exception
+func allPartFileExist(dir string, fileName string, totalSubtrieFiles int) error {
```

Comment:
It's a bit confusing to use `totalParts` as I consider the header file as one of the part file. However, the header file has different name pattern than others, since other files has `.xx` in the name, whereas the header file doesn't have it. 

That's why I think it's easier to let this function decide the total parts